From 1cc85be23facf9ff44a5a190d689d44b0087c407 Mon Sep 17 00:00:00 2001
From: chandavath <rcrajagopal@habana.ai>
Date: Wed, 12 Jun 2024 21:27:16 +0300
Subject: [PATCH] [SW-188535] g3: security: ported function from preboot rtos

Ported security function required to validate svn and validate
public key & index from preboot rtos.

Moved security code common between Gaudi2 and Gaudi3 into
lib/common/security.c to reuse in both places

Change-Id: I7d4478ad4444d730f09f681ae0a87a93eb5071c8
Signed-off-by: chandavath <rcrajagopal@habana.ai>
---

diff --git a/app/ext_rom/Makefile b/app/ext_rom/Makefile
index 0cc8996..d9aed2e 100644
--- a/app/ext_rom/Makefile
+++ b/app/ext_rom/Makefile
@@ -69,6 +69,7 @@
 	$(PREBOOT_ROOT_DIR)/lib/common/module_pci.o \
 	$(PREBOOT_ROOT_DIR)/lib/common/boot_info.o \
 	$(PREBOOT_ROOT_DIR)/lib/common/crc32.o \
+	$(PREBOOT_ROOT_DIR)/lib/common/security.o \
 	$(PREBOOT_ROOT_DIR)/lib/preboot/fit_regions.o \
 	$(PREBOOT_ROOT_DIR)/lib/preboot/module_fwupd.o \
 	$(PREBOOT_ROOT_DIR)/lib/preboot/module_uart.o \
diff --git a/arch/gaudi2/common/security.c b/arch/gaudi2/common/security.c
index 41adb3f..b1cc8ce 100644
--- a/arch/gaudi2/common/security.c
+++ b/arch/gaudi2/common/security.c
@@ -548,7 +548,7 @@
 	return false;
 }
 
-static int preboot_security_get_minimal_svn(void)
+int preboot_security_get_minimal_svn(void)
 {
 	const struct hl_sec_efuse *fuse;
 	const u32 *svn_bits;
@@ -597,21 +597,6 @@
 
 	preboot_puts_hex("SVN on eFuse: ", svn, 0);
 	return svn;
-}
-
-bool preboot_security_svn_is_valid(int sec_ver)
-{
-	int min_svn = preboot_security_get_minimal_svn();
-
-	if (min_svn < 0)
-		return false;
-
-	if (sec_ver < min_svn || sec_ver > SVN_MAX_VERSIONS) {
-		preboot_puts_hex("Illegal security version: ", sec_ver, false);
-		return false;
-	}
-
-	return true;
 }
 
 bool preboot_security_key_is_valid(int key_idx, const void *pub_key,
@@ -663,94 +648,6 @@
 		hash_compare_len = SHA256_SUM_LEN;
 
 	return !memcmp(&otp->keys[key_idx * SHA_SUM_LEN], value, hash_compare_len);
-}
-
-static struct hl_fit_sec_desc fit_sec;
-static u8 signature[FIT_SIGNATURE_SIZE];
-
-int preboot_verify_img_sig(const void *img, size_t img_size, enum fit_sec_failure *res,
-				u32 *key_idx)
-{
-	const void *fit_sec_off;
-	const void *pub_key_fit;
-
-	if (!preboot_security_is_enabled())
-		return 0;
-
-	/* hl fit files are padded so the size is 4 bytes aligned */
-	img_size = ROUND_UP4(img_size);
-
-	/* security descriptor is at the footer of the image */
-	fit_sec_off = (const void *)((u8 *)img + img_size);
-
-	/* check that the calculated memory range are legal */
-	if (!preboot_sram_is_valid_address_range((uintptr_t)fit_sec_off,
-						 sizeof(fit_sec))) {
-		preboot_puts("Invalid address range\n");
-		return 1;
-	}
-
-	memcpy(&fit_sec, fit_sec_off, FIT_SEC_DESC_SIZE);
-
-	if (fit_sec.magic_number != FIT_SEC_MAGIC) {
-		preboot_puts("Bad magic, security descriptor not found\n");
-		return 1;
-	}
-
-	if ((fit_sec.desc_version_maj != FIT_SEC_DESC_VER_MAJOR) ||
-	    (fit_sec.desc_version_min != FIT_SEC_DESC_VER_MINOR)) {
-		preboot_puts("security descriptor version mismatch\n");
-		return 1;
-	}
-
-	if (fit_sec.desc_size < FIT_SEC_DESC_SIZE) {
-		preboot_puts("security descriptor size too small\n");
-		return 1;
-	}
-
-	if (fit_sec.img_size != img_size) {
-		preboot_puts("image size mismatch with security descriptor\n");
-		return 1;
-	}
-
-	if (!preboot_security_svn_is_valid(fit_sec.sec_version)) {
-		preboot_puts("security descriptor svn not valid\n");
-		*res = FIT_SEC_SVN_INVALID;
-		return 1;
-	}
-
-	if (preboot_security_key_is_revoked(fit_sec.key_num)) {
-		preboot_puts_hex("Key revoked: ", fit_sec.key_num, false);
-		*res = FIT_SEC_KEY_REVOKED;
-		return 1;
-	}
-
-	pub_key_fit = &fit_sec.pub_key;
-
-	if (!preboot_security_key_is_valid(fit_sec.key_num, pub_key_fit,
-					   FIT_PUB_KEY_SIZE, NULL, 0)) {
-		preboot_puts("Key validation failed\n");
-		*res = FIT_SEC_SIGN_FAILED;
-		return 1;
-	}
-
-	/*
-	 * signature is calculated on both the fit image and the security
-	 * descriptor with the signature field of the descriptor zeroized.
-	 */
-	memcpy(signature, fit_sec.signature, FIT_SIGNATURE_SIZE);
-	memset(fit_sec.signature, 0, FIT_SIGNATURE_SIZE);
-
-	if (fit_verify_img_sig(img, img_size, &fit_sec, FIT_SEC_DESC_SIZE,
-			       pub_key_fit, signature, FIT_SIGNATURE_SIZE)) {
-		preboot_puts("image signature verification failed\n");
-		*res = FIT_SEC_SIGN_FAILED;
-		return 1;
-	}
-
-	*key_idx = fit_sec.key_num;
-
-	return 0;
 }
 
 #define EN_ARC_SEC_PRIV				0x3
diff --git a/arch/gaudi3/common/security.c b/arch/gaudi3/common/security.c
index 790f0c8..b9c9eb6 100644
--- a/arch/gaudi3/common/security.c
+++ b/arch/gaudi3/common/security.c
@@ -33,6 +33,12 @@
 #include <hl_protbits_fw_if.h>
 #include <security.h>
 #include "security_defs.h"
+#include <hl_boot_desc.h>
+#include <boot_info.h>
+#include <arcboot_export_dev.h>
+#include <fit.h>
+#include <fit_rsa.h>
+#include <sha.h>
 
 #define RR_HBW_PRIV_RANGE_MIN		0x0
 /* Below value is to keep HBW RRs configured at 16KB granularity */
@@ -1078,3 +1084,90 @@
 		preboot_puts("Security disabled.");
 	}
 }
+
+void preboot_security_boot_info_write_key_idx(u32 *key_idx)
+{
+	if (hl_boot_info_set(HL_BOOT_INFO_KEY_IDX, key_idx, sizeof(u32)))
+		preboot_puts("BOOT_INFO: Failed to set fw info\n");
+}
+
+bool preboot_security_is_enabled(void)
+{
+	return !!(sec_data.efuse_sys_cfg & EFUSE_SEC_EN_MASK);
+}
+
+bool preboot_img_validation_is_enabled(void)
+{
+	return !!(sec_data.efuse_sys_cfg & EFUSE_SEC_SIG_VAL_EN_MASK);
+}
+
+int preboot_security_get_svn_ver(void)
+{
+	return sec_data.sec_ver;
+}
+
+int preboot_security_get_minimal_svn(void)
+{
+	return preboot_security_get_svn_ver();
+}
+
+bool hl_extract_hash_from_key(const void *pub_key, u8 *value)
+{
+	struct rsa_key_prop prop;
+	u8 key_props[PUB_KEY_SZ];
+	const char *key_name = HL_SEC_KEY_NAME;
+
+	/*
+	 * ROM stores hash of public key properties. The public
+	 * key received as part of fit is the complete public key dtb of
+	 * 3144-bytes. Extract key properties and compute hash on 784 bytes
+	 * of key properties to compare with stored ROM/eFUSE hash.
+	 */
+	if (fit_get_key_prop(pub_key, key_name, &prop)) {
+		preboot_puts("Failed to get key properties\n");
+		return false;
+	}
+
+	memset(key_props, 0, sizeof(key_props));
+	memcpy(key_props, prop.rr, KEY_PROP_RR_SZ);
+	memcpy((key_props + KEY_PROP_RR_SZ), prop.modulus, KEY_PROP_MOD_SZ);
+	memcpy((key_props + KEY_PROP_RR_SZ + KEY_PROP_MOD_SZ), prop.public_exponent,
+		KEY_PROP_EXP_SZ);
+	memcpy((key_props + KEY_PROP_RR_SZ + KEY_PROP_MOD_SZ + KEY_PROP_EXP_SZ),
+		&prop.n0inv, KEY_PROP_INV_SZ);
+	memcpy((key_props + KEY_PROP_RR_SZ + KEY_PROP_MOD_SZ + KEY_PROP_EXP_SZ +
+		KEY_PROP_INV_SZ), &prop.num_bits, KEY_PROP_NB_SZ);
+
+	sha_csum(key_props, PUB_KEY_SZ, value);
+
+	return true;
+}
+
+bool preboot_security_key_is_valid(int key_idx, const void *pub_key,
+			 int pub_key_len, const void *pub_key_ext,
+			 int pub_key_ext_len)
+{
+	u8 value[SHA_SUM_LEN];
+	uint8_t expected_hash[SHA384_SUM_LEN];
+
+	if (!hl_extract_hash_from_key(pub_key, value))
+		return false;
+
+	if (security_rom_get_key_hash_by_idx((uint32_t *)expected_hash, key_idx))
+		return false;
+
+	return !memcmp(value, expected_hash, SHA384_SUM_LEN);
+}
+
+bool preboot_security_key_is_revoked(int key_idx)
+{
+	if (key_idx < 0 || key_idx >= SEC_KEY_HASH_COUNT_IN_ROM)
+		return true;
+
+	if (sec_data.key_revoke & BIT(key_idx))
+		return true;
+
+	fw_info.sec_revoked_keys = sec_data.key_revoke;
+
+	return false;
+}
diff --git a/include/security.h b/include/security.h
index 0412f7b..928f78d 100644
--- a/include/security.h
+++ b/include/security.h
@@ -34,5 +34,11 @@
 void preboot_security_sram_pci_set_ro(bool en);
 bool preboot_security_fuse_override(void);
 bool ppboot_sys_cfg_efuse_is_set(u32 mask);
+bool preboot_security_svn_is_valid(int sec_ver);
+bool preboot_security_key_is_valid(int key_idx, const void *pub_key,
+			 int pub_key_len, const void *pub_key_ext,
+			 int pub_key_ext_len);
+bool preboot_security_key_is_revoked(int key_idx);
+int preboot_security_get_minimal_svn(void);
 
 #endif /* _HL_SECURITY_H_ */
diff --git a/lib/common/security.c b/lib/common/security.c
new file mode 100644
index 0000000..e0bcff5
--- /dev/null
+++ b/lib/common/security.c
@@ -0,0 +1,118 @@
+// SPDX-License-Identifier:     GPL-2.0+
+/*
+ *
+ * (C) Copyright 2024 HabanaLabs, Ltd.
+ *
+ */
+#include <baremetal.h>
+#include <hl_fit_if.h>
+#include <hl_sec_if.h>
+
+#include <security.h>
+#include <preboot/fit.h>
+#include <string.h>
+#include <uart.h>
+
+static struct hl_fit_sec_desc fit_sec;
+static u8 signature[FIT_SIGNATURE_SIZE];
+
+
+bool preboot_security_svn_is_valid(int sec_ver)
+{
+	int min_svn = preboot_security_get_minimal_svn();
+
+	if (min_svn < 0)
+		return false;
+
+	if (sec_ver < min_svn || sec_ver > SVN_MAX_VERSIONS) {
+		preboot_puts_hex("Illegal security version: ", sec_ver, false);
+		return false;
+	}
+
+	return true;
+}
+
+int preboot_verify_img_sig(const void *img, size_t img_size, enum fit_sec_failure *res,
+				u32 *key_idx)
+{
+	const void *fit_sec_off;
+	const void *pub_key_fit;
+
+	if (!preboot_security_is_enabled())
+		return 0;
+
+	/* hl fit files are padded so the size is 4 bytes aligned */
+	img_size = ROUND_UP4(img_size);
+
+	/* security descriptor is at the footer of the image */
+	fit_sec_off = (const void *)((u8 *)img + img_size);
+
+	/* check that the calculated memory range are legal */
+	if (!preboot_sram_is_valid_address_range((uintptr_t)fit_sec_off,
+						 sizeof(fit_sec))) {
+		preboot_puts("Invalid address range\n");
+		return 1;
+	}
+
+	memcpy(&fit_sec, fit_sec_off, FIT_SEC_DESC_SIZE);
+
+	if (fit_sec.magic_number != FIT_SEC_MAGIC) {
+		preboot_puts("Bad magic, security descriptor not found\n");
+		return 1;
+	}
+
+	if ((fit_sec.desc_version_maj != FIT_SEC_DESC_VER_MAJOR) ||
+	    (fit_sec.desc_version_min != FIT_SEC_DESC_VER_MINOR)) {
+		preboot_puts("security descriptor version mismatch\n");
+		return 1;
+	}
+
+	if (fit_sec.desc_size < FIT_SEC_DESC_SIZE) {
+		preboot_puts("security descriptor size too small\n");
+		return 1;
+	}
+
+	if (fit_sec.img_size != img_size) {
+		preboot_puts("image size mismatch with security descriptor\n");
+		return 1;
+	}
+
+	if (!preboot_security_svn_is_valid(fit_sec.sec_version)) {
+		preboot_puts("security descriptor svn not valid\n");
+		*res = FIT_SEC_SVN_INVALID;
+		return 1;
+	}
+
+	if (preboot_security_key_is_revoked(fit_sec.key_num)) {
+		preboot_puts_hex("Key revoked: ", fit_sec.key_num, false);
+		*res = FIT_SEC_KEY_REVOKED;
+		return 1;
+	}
+
+	pub_key_fit = &fit_sec.pub_key;
+
+	if (!preboot_security_key_is_valid(fit_sec.key_num, pub_key_fit,
+					   FIT_PUB_KEY_SIZE, NULL, 0)) {
+		preboot_puts("Key validation failed\n");
+		*res = FIT_SEC_SIGN_FAILED;
+		return 1;
+	}
+
+	/*
+	 * signature is calculated on both the fit image and the security
+	 * descriptor with the signature field of the descriptor zeroized.
+	 */
+	memcpy(signature, fit_sec.signature, FIT_SIGNATURE_SIZE);
+	memset(fit_sec.signature, 0, FIT_SIGNATURE_SIZE);
+
+	if (fit_verify_img_sig(img, img_size, &fit_sec, FIT_SEC_DESC_SIZE,
+			       pub_key_fit, signature, FIT_SIGNATURE_SIZE)) {
+		preboot_puts("image signature verification failed\n");
+		*res = FIT_SEC_SIGN_FAILED;
+		return 1;
+	}
+
+	*key_idx = fit_sec.key_num;
+
+	return 0;
+}
